readme.md

![Zelda: Mystery of Solarus DX Logo](/data/logos/logo.png)

# Zelda: Mystery of Solarus XD

This package contains the data files of the game Zelda: Mystery of Solarus DX.
This quest is a free, open-source game that works with [Solarus](https://github.com/christopho/solarus),
an open-source Zelda-like 2D game engine.
To play this game, you need Solarus.
We assume in this readme that Solarus is already installed.

See our [development blog](http://www.solarus-games.org) to get more
information and documentation about Solarus and our games.

## Table of Contents
- [1. Play directly](#1-play-directly)
- [2. Install the quest](#2-install-the-quest)
	- [2.1. Default settings](#21-default-settings)
	- [2.2. Change the install directory](#22-change-the-install-directory)


## 1.  Play directly

You need to specify to the `solarus_run` binary the path of the quest data files
to use. The binary `solarus_run` accepts two forms of quest paths:
 - a directory having a subdirectory named `data` with all data inside,
 - a directory having a zip archive `data.solarus` with all data inside.

Thus, to run zsdx, if the current directory is the one that
contains the `data` subdirectory (and this readme), you can type:
```bash
$ solarus_run .
```
Or without arguments, if Solarus was compiled with the default quest set to `"."`:
```bash
$ solarus_run
```


## 2. Install the quest

###  2.1. Default settings

If you want to install zsdx, cmake and zip are recommended.
Just type:
```bash
$ cmake .
$ make
```

This generates the `data.solarus` archive that contains all data files
of the quest. You can then install it with
```bash
$ make install
```

This installs the following files (assuming that the install directory
is `/usr/local`):
- the quest data archive (`data.solarus`) in `/usr/local/share/solarus/zsdx/`
- a script called `zsdx` in `/usr/local/bin/`

The `zsdx` script launches `solarus_run` with the appropriate command-line argument
to specify the quest path.
This means that you can launch the zsdx quest with the command:
```bash
$ zsdx
```
which is equivalent to:
```bash
$ solarus_run /usr/local/share/solarus/zsdx
```


### 2.2. Change the install directory

You may want to install zsdx in another directory
(e.g. so that no root access is necessary).
You can specify this directory as a parameter of cmake:
```bash
$ cmake -D CMAKE_INSTALL_PREFIX=/home/your_directory .
$ make
$ make install
```
This installs the files described above, with the
`/usr/local` prefix replaced by the one you specified.
The script generated runs `solarus_run` with the appropriate quest path.


ChangeLog
zsdx 1.10.3 (2015-12-20)

* Fix Italian dialogs exceeding the dialog box (thanks Marco).
* Fix game-over stopped sometimes using a workaround (#87).

zsdx 1.10.2 (2015-11-22)

* Upgrade to Solarus 1.4.5.
* Add Italian translation (thanks Marco).
* Make fairies play a sound when restoring life even when already full (#101).
* Fix dialog cursor drawn at wrong position after successive questions (#105).
* Dungeon 1: fix unintentional extra difficulty with block alignment (#102).
* Dungeon 3: fix enemy entering mini-boss room (#107).
* Dungeon 9: fix spikes sending back to wrong room (#108).
* Dungeon 10: fix evil tiles door not opening sometimes (#94).
* End screen: fix freezing the hero (#109).

zsdx 1.10.1 (2015-05-09)

* Upgrade to Solarus 1.4.1.

zsdx 1.10.0 (2015-05-02)

* Upgrade to Solarus 1.4.0.
* Fix a minor tile issue in dungeon 8 entrance.
* Fix the feather dialog that could be skipped in some languages.
* Fix empty line in English dialog describing a bottle with water.
* Fix empty buttons in savegame menu when changing language (#90).
* Fix translation errors of two items in German (#98).
* Dungeon 5: fix a missing door on the minimap (#99).
* Dungeon 9: don't allow to skip the boss hint dialog (#93).

zsdx 1.9.0 (2014-08-21)

* Use Solarus 1.3.0.
* New world minimap. By Neovyse.
* Make cmake paths more modifiable. By hasufell.
* Fix direction of vertical movement on joypad in menus (#85). By xethm55.
* Clarify license of some files.

zsdx 1.8.0 (2014-05-06)

* Use solarus 1.2.0.
* Replace fixed8.fon by a TTF font due to an SDL2 regression (#59).
* Switch fullscreen also with F11.
* Dungeon 3 5F: fix falling too early in the hole due to incorrect layers.

zsdx-1.7.1 (2013-12-01)

* Use Solarus 1.1.1.
* Fix the heart meter not updated until the first save/reload (#67).
* Fix wrong teletransportation in Billy's cave after falling in a hole (#66).
* Fix the debugging console no longer working after using F1, F2 or F3.

zsdx-1.7.0 (2013-10-13)

* Use solarus 1.1.0.
* Add simplified Chinese and traditional Chinese translations (beta).
* Replace .spc musics by .it ones (much faster) (#17).
* Add an animated Solarus logo from Maxs (#57).
* Fix savegames freshly created with Solarus 0.9 and never run yet.
* Fix a slight alignment issue with the hurt animation of the hero.
* Fix a small breach in dungeon 9 4F in the timed chest room (#4).
* Fix easy infinite rupees in the waterfall cave (#13).
* Fix low health beeping playing at final screen (#56).
* Add a teletransporter from south lake to the cave under the waterfall (#10).
* Dungeon 7 boss: the player could get stuck in the boss room (#6).
* Dungeon 7 boss: change the misleading hurt sound of the tail (#7).
* Dungeon 7: help the player be aligned correctly to obtain the boss key (#8).
* Dungeon 3 2F: fix two switches that disappeared when activating them.
* Dungeon 2 boss: fix tile disappearing issue (#14).
* Dungeon 9 2F: fix a minor graphical issue.

zsdx-1.6.2 (2013-06-25)

* Use solarus 1.0.4.
* Fix line errors in English dialogs.

zsdx-1.6.1 (2013-06-25)

* Use solarus 1.0.3.

zsdx-1.6 (2013-06-22)

* Use solarus 1.0.2.
* Improve the English translation.
* New title screen from Neovyse (#45).
* Fix a typo in French dialogs.
* Fix a door bug that could get the hero stuck in empty rooms.
* Fix saving the state of Billy's door.
* Dungeon 1: keep the final door open when coming back in the boss room.
* Dungeon 7 2F: save the north-west torches.
* Savegames menu: the joypad was not working to delete a savegame.
* Dungeon 10 B1: fix a cauldron the hero could walk on (#20).
* Rupee house: don't let the player to get the piece of heart with the sword.
* Dungeon 6 2F: move crystal blocks to avoid a possible breach.
* Outside world B3: fix a minor tile error (#12).
* Dungeon 8 B1: fix a minor tile error (#2).
* Dungeon 8 B3: fix a minor tile error.

zsdx-1.5.2 (2013-05-08)

* Use solarus 0.9.3.
* German dialogs available.
* Castle 1F: fix NPCs possibly overlapping the hero.
* West mountains cave: fix a chest possibly overlapping the hero.
* Fix wrong displaying of a question in French dialogs.
* Fix three typos in French dialogs (thanks Stella).
* Improve the English translation (thanks Rypervenche).
* Fix a wrong empty line in English dialogs.
* Dungeon 5: fix an error the Spanish translation (thanks Joan).
* Fix the CMake script that creates the zip archive (#1).

zsdx-1.5.1 (2012-05-20)

* Spanish dialogs available (beta)
* English dialogs: fix dialogs with questions
* English dialogs: various improvements
* Dungeon 10: the hero could land in a wall when falling into a hole
* Dungeon 10: help the player a little bit more with the weak wall
* Dungeon 10: change the location where a key appears
* Dungeon 10: fix minor mapping errors
* Dungeons 7 and 10: fix blue flames in the boss
* Dungeon 5 B1: give bombs to make sure the hero can pass the weak walls
* Dungeon 5 2F: move the hint stone to the room of the corresponding puzzle
* Dungeon 5 2F: two enemies could be stopped by strange invisible walls
* Dungeon 6 3F: really fix the breach
* Savegame variable 0 was used twice (no consequence)
* Always set the variant of _none and _random treasures to 1

zsdx-1.5.0 (2012-04-03)

* Use solarus 0.9.2
* English translation available
* Dungeon 5: the piece of heart could be obtained very easily with the sword
* Dungeon 5: a black guard was never seeing the hero

zsdx-1.4.1 (2012-02-18)

* Dungeon 5: fix wrong coordinates of two blocks saved

zsdx-1.4 (2012-02-12)

* Use solarus-0.9.1
* The Pegasus Shoes can now be used with the action key
* New end screen
* Dungeon 1: restart the boss minillosaurs after they are hurt
* Dungeon 9: the player could be stuck if using a bomb on a crystal
* Dungeon 9: the player could be stuck because of an incorrect jump length
* Dungeon 7: the player could be stuck in a room if he had no key
* West mountains cave 1F: the player could be stuck if a door was closed

data/maps/3.lua
local map = ...
local game = map:get_game()

-- Outside world A3.

local monkey_sprite = nil
local monkey_jumps = 0
local monkey_jump_speed = 48

local function random_walk(npc)

  local m = sol.movement.create("random_path")
  m:set_speed(32)
  m:start(npc)
end

local function remove_dungeon_2_door()
  dungeon_2_door:remove()
  dungeon_2_door_tile:set_enabled(false)
end

-- Function called when the map starts.
-- The NPCs are initialized.
function map:on_started(destination)

  -- enable dark world
  if game:get_value("b905") then
    sol.audio.play_music("dark_world")
    map:set_tileset(13)
  else
    sol.audio.play_music("overworld")
  end

  if game:get_value("b24") then
    -- remove the monkey from Link's house entrance
    monkey:remove()
  else
    monkey_sprite = monkey:get_sprite()
  end

  if game:get_value("b89") then
    -- remove the dungeon 2 door
    remove_dungeon_2_door()
  end

  -- initialize NPCs
  random_walk(hat_man)
  random_walk(grand_son)

  if game:is_dungeon_finished(1) then
    how_to_save_npc:remove()
  else
    random_walk(how_to_save_npc)
  end

  -- smith cave with thiefs
  if game:get_value("b155") and not game:get_value("b156") then
    to_smith_cave:set_enabled(false)
  else
    to_smith_cave_thiefs:set_enabled(false)
  end

  -- entrances of houses
  local entrance_names = {
    "sahasrahla", "shop", "cake_shop", "hero", "telepathic_booth"
  }
  for _, entrance_name in ipairs(entrance_names) do
    local sensor = map:get_entity(entrance_name .. "_door_sensor")
    local tile = map:get_entity(entrance_name .. "_door")

    sensor.on_activated_repeat = function()
      if hero:get_direction() == 1
        and tile:is_enabled() then
        tile:set_enabled(false)
        sol.audio.play_sound("door_open")
      end
    end
  end
end

function monkey:on_interaction()

  if not game:get_value("b24") then
    -- monkey first dialog
    sol.audio.play_sound("monkey")
    game:start_dialog("outside_world.village.monkey", function()

      -- show another message depending on the shield
      if not game:has_ability("shield") then
	game:start_dialog("outside_world.village.monkey.without_shield")
      else
	game:start_dialog("outside_world.village.monkey.with_shield", function()
	  -- make the monkey leave
	  hero:freeze()
	  sol.audio.play_sound("monkey")
	  local m = sol.movement.create("jump")
	  m:set_direction8(1)
	  m:set_distance(24)
	  m:set_ignore_obstacles(true)
	  m:set_speed(monkey_jump_speed)
          m:start(monkey)
	  monkey_sprite:set_animation("jumping")
	  monkey_jumps = 1
	  game:set_value("b24", true)
	end)
      end
    end)
  else
    sol.audio.play_sound("monkey")
    game:start_dialog("outside_world.dungeon_2_entrance.monkey")
  end
end

function dungeon_2_door:on_interaction()

  -- open the door if the player has the Rock Key
  if game:has_item("rock_key") then
    sol.audio.play_sound("door_open")
    sol.audio.play_sound("secret")
    game:set_value("b89", true)
    remove_dungeon_2_door()
  else
    game:start_dialog("outside_world.rock_key_required")
  end
end

function hat_man:on_interaction()

  if game:is_dungeon_finished(1) then
    game:start_dialog("outside_world.village.hat_man_npc_waterfall")
  else
    game:start_dialog("outside_world.village.hat_man_npc")
  end
end

function tree_woman:on_interaction()

  game:start_dialog("outside_world.village.tree_woman", function(answer)
    if answer ~= "skipped" then
      hero:start_treasure("rupee", 1)
    end
  end)
end

-- Function called when the monkey has finished jumping
function monkey:on_movement_finished()

  if monkey_jumps == 1 then
    -- first jump finished: wait a little amount of time before jumping again
    sol.timer.start(300, function()
      -- start the second jump
      sol.audio.play_sound("monkey")
      local m = sol.movement.create("jump")
      m:set_direction8(2)
      m:set_distance(56)
      m:set_ignore_obstacles(true)
      m:set_speed(monkey_jump_speed)
      m:start(monkey)
      monkey_sprite:set_animation("jumping")
      monkey_jumps = 2
    end)
  elseif monkey_jumps == 2 then
    -- second jump finished: start the last jump
    sol.audio.play_sound("monkey")
    local m = sol.movement.create("jump")
    m:set_direction8(1)
    m:set_distance(64)
    m:set_ignore_obstacles(true)
    m:set_speed(monkey_jump_speed)
    m:start(monkey)
    monkey_sprite:set_animation("jumping")
    monkey_jumps = 3
  else
    -- last jump finished: remove the monkey from the map and unfreeze the hero
    monkey:remove()
    hero:unfreeze()
  end
end

function dungeon_3_entrance_weak_block:on_opened()
  sol.audio.play_sound("secret") -- play the sound only once
end

function diagonal_jump_sensor:on_activated()

  -- We don't use a jumper here because we don't want the delay.
  sol.audio.play_sound("jump")
  hero:start_jumping(5, 56, true)
end

